New tutorial for ROS 2 launch files (#353)
ros2/ros2_documentation: source/Tutorials/LaunchFiles/CreatingLaunchFiles.rst

.. _ROS2Launch:

Creating a launch file
======================

**Goal:** Create a launch file to run a complex ROS 2 system.

**Tutorial level:** Beginner

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

A launch file can be thought of as an entry point to a ROS 2 system.

Launch files allow you to start up and configure a number of executables containing ROS 2 nodes simultaneously.

In the tutorials up until now, you have been opening new terminals for every new node you run.

> As you create more complex systems with more and more nodes running simultaneously, opening terminals and reentering configuration details becomes tedious.

Reviewer: @maryaB-osr +1 to the concise explanation, I would only reverse the order of the paragraphs:

```
In the tutorials up until now, you have been opening new terminals for every new node you run.
As you create more complex systems with more and more nodes running simultaneously, opening terminals and reentering configuration details becomes tedious.

Launch files allow you to start up and configure a number of executables containing ROS 2 nodes simultaneously.

*In a sense*, a launch file can be thought of as an entry point to a ROS 2 system.
```

Reviewer: CC @wjwwood for valuable feedback.

Reviewer 2: I think reversing the order makes sense as well.

----

I think the line:

> *In a sense*, a launch file can be thought of as an entry point to a ROS 2 system.

Might be unnecessary, or instead maybe it should read something like this:

> Often, the recommended way to use a piece of software in ROS is to run one of the associated launch file, as it allows the original developers to put more default configurations in place than would be comfortable with a single executable.

My main complaint is that "entry point" is a bit vague in this context, and so my suggestion is trying to say a similar thing but with plainer words? Not sure though.

Prerequisites
-------------

This tutorial uses the ``rqt_graph`` and ``turtlesim`` packages.

.. link to turtlesim + rqt tutorial

You will also need to use a text editor of your preference.

As always, don’t forget to source ROS 2 in **every new terminal you open**.

Tasks
-----

1 Setup
^^^^^^^

Create a new directory to store your launch file:

.. code-block:: bash

  mkdir launch

Create a launch file named ``turtlesim_mimic_launch.py`` by entering the following command in the terminal:

.. tabs::

   .. group-tab:: Linux

      .. code-block:: bash

        touch turtlesim_mimic_launch.py

   .. group-tab:: macOS

      .. code-block:: bash

        touch turtlesim_mimic_launch.py

   .. group-tab:: Windows

      .. code-block:: bash

        type nul > turtlesim_mimic_launch.py

You can also go into your system’s file directory using the GUI and create a new file that way.

Open the new file in your preferred text editor.

2 Write the launch file
^^^^^^^^^^^^^^^^^^^^^^^

Let’s put together a ROS 2 launch file using the ``turtlesim`` package and its executables.

> At the top of your new ``turtlesim_mimic_launch.py`` file, add the lines:

Reviewer 3: This is the first PR I'm reviewing with actual code, instead of just commands.  So I don't know what we want our convention for code examples to be.  But I think that people will appreciate all of the code up-front so that they can cut-n-paste the entire example.  The text that follows can then break down individual sections of code, explaining what they do.  See https://clalancette.github.io/pycdlib/example-creating-new-basic-iso.html where I for an example where put this idea into practice.

Also, line numbers in the code block are useful so you can refer to specific lines of code from the text.

Author: code up-front: ✔

line numbers: for some reason the linter's not been accepting the `:linenos:` directive for `code-block`. I'll try again though

Author: ```
D000 Error in "code-block" directive: unknown option: "linenos".
```
For some reason it works in the [next tutorial](https://github.com/ros2/ros2_documentation/pull/362/files#diff-867ca1fb9f0f0b9b9938ed7275276d55R336) when its nested under `.. tabs:: > .. group-tab:: > .. code-block::` though

Reviewer 2: Not sure, make sure you're using spaces and not tabs and that you're using the right number of indented spaces?

Maybe you need to indent the following Python code a second time:

https://github.com/ros2/ros2_documentation/pull/353/files#diff-ebe5f45a06c85b302cce7f170b047089R85

The other place you use `:linenos:` is like that:

https://github.com/ros2/ros2_documentation/pull/353/files#diff-ebe5f45a06c85b302cce7f170b047089R118-R122

Author: > Maybe you need to indent the following Python code a second time:
>  
> The other place you use :linenos: is like that:

But both of those lines are throwing errors, extra indentation or not.
I tried adjusting the spaces before `:linenos:` (in the [sphinx docs](https://www.sphinx-doc.org/en/1.5/markup/code.html#line-numbers) it seems like 3 spaces?) but now the linter complaints are even more bizarre

Reviewer 2: Looks like @jacobperron had the same issue and suppressed the warning? https://github.com/ros2/ros2_documentation/pull/169#issuecomment-484721871

Author: I think D000 warnings can come up for a few different things, right? If we suppress those for all the tutorials, won't it be possible that some actual errors will get through? Because I need line numbers on a handful, not just this one.

Reviewer 2: Yeah, I don't know what we should do, but I was just pointing out that they've been ignored in the past. Sorry I don't have time right now to investigate the technical reason for the warning right now.

Author: @wjwwood That's fine, it's not necessary right now. Thanks for your input so far!

Reviewer 4: Correct, I'd added exceptions to the CI script of the turtorials under Tutorials/Actions as they make heavy use of `:linenos:`.

It's not the best solution, but I think it's okay if you want to add another exception for this specific tutorial file (not necessarily all of the tutorials).

Reviewer 4: @maryaB-osr I think the follow change to the [.travis.yml](https://github.com/ros2/ros2_documentation/blob/6441d343b574107c296505ab59335ecccc3e3b74/.travis.yml) file should do it:
```diff
-  - doc8 --ignore D001 --ignore-path build --ignore-path source/Tutorials/Actions
-  # ignore D000 in action tutorials to allow for :linenos:
-  - doc8 --ignore D000 --ignore D001 --ignore-path build source/Tutorials/Actions
+  - doc8 --ignore D001 --ignore-path build --ignore-path source/Tutorials/Actions --ignore-path source/Tutorials/LaunchFiles/CreatingLaunchFiles.rst
+  # ignore D000 in files that use :linenos:
+  - doc8 --ignore D000 --ignore D001 --ignore-path build source/Tutorials/Actions source/Tutorials/LaunchFiles/CreatingLaunchFiles.rst
```

.. code-block:: python

    from launch import LaunchDescription
    from launch_ros.actions import Node

> These import statements pull in the Python modules necessary for launch files.

Reviewer: @maryaB-osr nit: consider 
```suggestion
These import statements pull in the some Python ``launch`` modules.
```

The only one recurring import is `from launch import LaunchDescription`, the rest depends on the specific actions being used.

Next, begin the launch description itself by adding the lines:

.. code-block:: python

  def generate_launch_description():
      return LaunchDescription([

      ])

Let's create a system of three nodes, all from the ``turtlesim`` package.
The goal of the system is to launch two turtlesim windows, and have one turtle mimic the movements of the other.

> To launch two turtlesim windows with your launch file, add the following code between the brackets of the ``return`` statement:

Reviewer: @maryaB-osr consider: 
```suggestion
To launch two turtlesim windows with your launch file, add the following actions to your launch description as follows;
```
and right after show the `Node` instantiations in their context, as in: 

```python
return LaunchDescription([
    Node(), ...
])
```
IMO it's much more explicit this way. WDYT?

Author: Much better, thanks :)

Author: Actually, now that I've gotten to Chris's recommendation to put the full launch file first and then dissect it, I'm going to keep it as just the nodes. Otherwise this would've been better

.. code-block:: python

  Node(
      package='turtlesim', node_namespace='turtlesim1',
      node_executable='turtlesim_node', node_name='sim',
      output='screen'
      ),
  Node(
      package='turtlesim', node_namespace='turtlesim2',
      node_executable='turtlesim_node', node_name='sim',
      output='screen'

>       ),

Reviewer: @maryaB-osr watch out for indentation.

Reviewer 3: I'd also suggest putting one keyword per line, which:

1)  Makes it clear at a glance how many keywords there are, and
1)  Allows you to refer to them by line number more easily (once we have line numbers)

Note the only difference between the two nodes is their ``node_namespace`` values.

> Unique namespaces allow the system to start two simulators without having name conflicts.

Reviewer: @maryaB-osr consider 
```suggestion
Unique namespaces allow the system to start two simulators without node names' nor topic names' conflicts.
```

> Without unique namespaces, the nodes would clash when trying to publish on the same topic with the same name.

Reviewer: @maryaB-osr thinking this over, instead of talking about clashes we could state that there'd be no way to address each turtle separately i.e. both receive commands over the same topic, both publish pose over the same topic.

Author: @hidmic So I changed the last line to this:
```
Both turtles in this system receive commands over the same topic and publish their pose over the same topic.
Without unique namespaces, there would be no way to distinguish between messages meant for one turtle or the other.
```
But now I think it's repetitive because it just seems like a long way of say of saying the previous sentence:
```
Unique namespaces allow the system to start two simulators without node name nor topic name conflicts.
```
Thoughts?

The final node is also from the ``turtlesim`` package, but a different executable: ``mimic``.

Add the following code below the previous nodes:

.. code-block:: python

  Node(
      package='turtlesim', node_executable='mimic', node_name='mimic',
      remappings=[
                  ('/input/pose', '/turtlesim1/turtle1/pose'),

>                   ('/output/cmd_vel', '/turtlesim2/turtle1/cmd_vel'),

Reviewer: @sloretz @wjwwood this is something I wanted to talk to you about.

The original roslaunch XML file would remap parent namespaces as in `/input:=/turtlesim1` but that doesn't seem to work on Dashing nor Eloquent. Globbing as in `/input/**:=/turtlesim1/\1` didn't work either. That's why I had to be super specific. Is that expected behavior?

Author: Let me know if/when this should be updated

Reviewer 2: > The original roslaunch XML file would remap parent namespaces as in `/input:=/turtlesim1` but that doesn't seem to work on Dashing nor Eloquent.

Can you link that? Maybe it's a use case we overlooked, but I think our remapping in ROS 2 has to be an exact match normally (no partial matches) on a topic/service/etc... name (namespaces are not remapped as far as I know).

> Globbing as in `/input/**:=/turtlesim1/\1` didn't work either. That's why I had to be super specific. Is that expected behavior?

I think that should have worked, @sloretz?

Reviewer 5: Whoops, missed this.
>> The original roslaunch XML file would remap parent namespaces as in /input:=/turtlesim1 but that doesn't seem to work on Dashing nor Eloquent.
>
> [...] I think our remapping in ROS 2 has to be an exact match normally (no partial matches) on a topic/service/etc... name (namespaces are not remapped as far as I know).

Yeah, it has to be an exact match. That case is supposed to be covered by `**`, but ...

>>    Globbing as in /input/**:=/turtlesim1/\1 didn't work either. That's why I had to be super specific. Is that expected behavior?
>
> I think that should have worked, @sloretz?

`**` `*`, and `\1` - `\9` were never implemented: ros2/rcl#233 ros2/rcl#232

                  ]

>       )

Reviewer: @maryaB-osr watch out for indentation.

This node has added configuration details in the form of remappings.

``mimic``'s ``/input/pose`` topic is remapped to ``/turtlesim1/turtle1/pose`` and it's ``/output/cmd_vel`` topic to ``/turtlesim2/turtle1/cmd_vel``.

> This means ``mimic`` will subscribe to ``turtlesim1/sim``'s position topic and republish it for ``turtlesim2/sim``'s movement topic to subscribe to.

Reviewer: @maryaB-osr watch out for leading slashes in node names i.e. `turtlesim1/sim` -> `/turtlesim1/sim` and `turtlesim2/sim` -> `/turtlesim2/sim`

Reviewer: Also, `position` -> `pose` and `movement` -> `velocity command`.

In other words, ``turtlesim2`` will mimic ``turtlesim1``'s movements.

> The complete ``turtlesim_mimic_launch.py`` file will look like this:

Reviewer 3: Oh, and here you have the complete example.  As I stated above, I would lead with this, followed by explanation; I think it will lead to less frustrated users.

.. code-block:: python

  from launch import LaunchDescription
  from launch_ros.actions import Node

  def generate_launch_description():
      return LaunchDescription([
              Node(
                      package='turtlesim', node_namespace='turtlesim1',
                      node_executable='turtlesim_node', node_name='sim',
                      output='screen'
              ),
              Node(
                      package='turtlesim', node_namespace='turtlesim2',
                      node_executable='turtlesim_node', node_name='sim',
                      output='screen'
              ),
              Node(
                      package='turtlesim', node_executable='mimic', node_name='mimic',
                      remappings=[
                          ('/input/pose', '/turtlesim1/turtle1/pose'),
                          ('/output/cmd_vel', '/turtlesim2/turtle1/cmd_vel'),
                      ]
              )
      ])

3 ros2 launch
^^^^^^^^^^^^^

To launch ``turtlesim_mimic_launch.py``, run the following command:

.. code-block::

  ros2 launch turtlesim_mimic_launch.py

.. note::

>   When launching a launch file from inside a package, the syntax is:

Reviewer: @maryaB-osr consider 
```suggestion
  When launching a launch file provided by a package, the syntax is:
```

  .. code-block::

>       ros2 launch <package_name> <launch_file_name>

Reviewer 3: I think this is a little confusing, hot on the heels of the command launching a file directly.  A bit more explanation would be helpful, something like:

```
It is possible to launch a launch file directly (as we do above), or provided by a package.  When it is provided by a package, the syntax is:

ros2 launch <package> <launch_file_name>
```

>   You will learn more about creating packages in a later tutorial

Reviewer 3: ```suggestion
  You will learn more about creating packages in a later tutorial.
```

.. todo: link to package creation tutorial

Two turtlesim windows will open, and you will see the following ``[INFO]`` messages telling you which nodes your launch file has started:

.. code-block::

  [INFO] [launch]: Default logging verbosity is set to INFO
  [INFO] [turtlesim_node-1]: process started with pid [11714]
  [INFO] [turtlesim_node-2]: process started with pid [11715]
  [INFO] [mimic-3]: process started with pid [11716]

To see the system in action, run the ``ros2 topic pub`` command on the ``/turtlesim1/turtle1/cmd_vel`` topic to get the first turtle moving:

.. code-block::

  ros2 topic pub -r 1 /turtlesim1/turtle1/cmd_vel geometry_msgs/msg/Twist '{linear: {x: 2.0, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: -1.8}}'

You will see both turtles following the same path.

.. image:: mimic.png

4 Introspect the system with rqt_graph
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

While the system is still running, you can open ``rqt_graph`` to get a better idea of the relationship between the nodes in your launch file.

> Run the command:

Reviewer 3: I think we need a new terminal here.

.. code-block::

  rqt_graph

.. image:: mimic_graph.png

A hidden node (the ``ros2 topic pub`` command you ran) is publishing data to the ``/turtlesim1/turtle1/cmd_vel`` topic on the left, which the ``/turtlesim1/sim`` node is subscribed to.

> The rest of the graph shows what was described earlier: ``mimic`` is subscribed to ``/turtlesim1/sim``'s pose topic, and publishes to ``/turtlesim2/sim``'s movement topic.

Reviewer: @maryaB-osr same as above on "movement topic".

^^^^^^^^^^^^^^^^^^^^^^^

Let’s put together a ROS 2 launch file using the ``turtlesim`` package and its executables.

Copy and paste the complete code into the ``turtlesim_mimic_launch.py`` file:

.. code-block:: python

    from  launch import LaunchDescription
    from launch_ros.actions import Node

        def generate_launch_description():
            return LaunchDescription([
                Node(
                    package='turtlesim',
                    node_namespace='turtlesim1',
                    node_executable='turtlesim_node',
                    node_name='sim'
                ),
                Node(
                    package='turtlesim',
                    node_namespace='turtlesim2',
                    node_executable='turtlesim_node',
                    node_name='sim'
                ),
                Node(
                    package='turtlesim',
                    node_executable='mimic',
                    node_name='mimic',
                    remappings=[
                        ('/input/pose', '/turtlesim1/turtle1/pose'),
                        ('/output/cmd_vel', '/turtlesim2/turtle1/cmd_vel'),
                    ]
                )

>             ])

Reviewer 6: For those who cut-n-paste this block of code, the extra indentation needs to be removed. Causes python interpreter to complain.

Author: @jackpien can you try it now?

Reviewer 6: Looks great. Thanks for quick fix.
